### PR TITLE
Extend cross-chain queries to support IBC pub/sub

### PR DESCRIPTION
**Pull request by angbrav**
This PR extends cross-chain queries to support IBC pub/sub (#759). The extensions are described in https://github.com/cosmos/ibc/issues/823.

### CODE REVIEW

File: spec/app/ics-031-crosschain-queries/README.md [hunk 1]
```diff
@@ -299,14 +359,27 @@ function CrossChainQueryResponse(
     // Delete the query from the local, private store.
     privateStore.delete(queryPath(queryId))
 
+    // Reassign query's heightUpdate and value in case the query is of type PERSISTENT and the
+    // result was SUCCESS
+    heightUpdate = null
+    value = data
+    if (query.type === PERSISTENT && data !== null) {
```
Comment (AdityaSripal) [inline spec/app/ics-031-crosschain-queries/README.md:366]: We also need to check that the previous value and height equal the previous resultRecord

File: spec/app/ics-031-crosschain-queries/README.md [hunk 2]
```diff
@@ -133,18 +154,24 @@ enum QueryResult {
...
+    udpateHeight: Height
     data: []byte
 }
 ```
 
 - The `id` field uniquely identifies the query at the querying chain.
+- The `type` field specifies the type of query.
+- The `path` field is the path queried at the queried chain.
 - The `result` field indicates whether the query was correctly executed at the queried chain and if the queried path exists.
+- The `udpateHeight` field indicates the height at which the queried chain updated the queried path with `data`.
```
Comment (AdityaSripal) [inline spec/app/ics-031-crosschain-queries/README.md:174]:
```suggestion
- The `updateHeight` field indicates the height at which the queried chain updated the queried path with `data`.
```

**Review by AdityaSripal — CHANGES_REQUESTED** [review]

**Comment by ethanfrey** [comment]
Maybe a bit out of scope, but I think important: how do we account for fees.

ICQ had some reward paid to the relayer IIRC. ICS29 allows for paying relayer fees on the packets we send.

If I make a subscription, and 1000s of packets come over, how do we account for fees? All free? Deduct from the subscriber for each contract? Have some "account" we need to fill up to pay for subscriptions and they stop when not paid???

I think we should consider relayer incentives as first class citizens in all new protocols we design.

**Comment by AdityaSripal** [comment]
I think the benefit of client queries is the efficiency. No need for packet roundtrips for each query. No need for having to send a packet to each individual querier. This makes a lot of sense for high demand queries (e.g. price feeds, oracle data, CNS, etc.)

However, client queries does introduce a brand new primitive that relayers and applications will have to adopt (building on top of the client layer instead of channels). We need to also rebuild certain concepts like relayer incentives on top as @ethanfrey mentioned.

I think with the right abstractions we can get the best of both worlds. Ideally we get the efficiency of client queries while still offering the same channel interface that relayers and applications are already used to. We then also get the higher-level primitives for free (ICS-29, any desired middleware).

My suggestion is to introduce a new channel type: `Broadcast channel`

It is effectively a one-to-many unidirectional channel.

This will allow a single write of `SendPacket` on a "publishing" chain to be received on an arbitrary number of subscribers.

It also preserves the same semantics that relayers and applications are already used to (though we get rid of acknowledgments and timeouts, the handshake will also be simplified)

Broadcast channels i think will also be a more broadly useful primitive beyond ibc queries

Will discuss the following with @angbrav  next week

**Comment by ethanfrey** [comment]
> My suggestion is to introduce a new channel type: Broadcast channel
>
> It is effectively a one-to-many unidirectional channel.
>
> This will allow a single write of SendPacket on a "publishing" chain to be received on an arbitrary number of subscribers.
>
> It also preserves the same semantics that relayers and applications are already used to (though we get rid of acknowledgments and timeouts, the handshake will also be simplified)
>
> Broadcast channels i think will also be a more broadly useful primitive beyond ibc queries

I agree with the direction of making this more generally useful primitive.

I am not sure we can hide it under the channel interface (which assumes a handshake... can we have 5 different handshakes on different connections to one channel?) and the fee middleware would be hard to extend as well.

I do agree it needs to be as generalised and reusable as the channel model to adopt into core ibc, allowing many different applications to be built using pub-sub, which we haven't yet envisioned, and showing how to add middleware-like functionality, routing etc, to pubsub.

Whether this is a new channel type. Or a new primitive on top of Connection that is the same level as channel (between Connection and app level) is an open question to me. But I agree strongly with the general direction you are pushing for.

File: spec/app/ics-031-crosschain-queries/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (angbrav) [reply spec/app/ics-031-crosschain-queries/README.md]: That's correct.

File: spec/app/ics-031-crosschain-queries/README.md [hunk 3]
```diff
@@ -316,6 +389,52 @@ function CrossChainQueryResponse(
...
+                                                  proof,
+                                                  resultRecord.path,
+                                                  data))
+    
+    resultRecord.updateHeight = linkedlistValue.height
+    resultRecord.data = linkedlistValue.value
+
+    // Store the result in the local, private store.
+    privateStore.set(resultQueryPath(queryIdentifier), resultRecord)
+
```
Comment (crodriguezvega) [inline spec/app/ics-031-crosschain-queries/README.md:428]: Remove empty line

```suggestion
```

**Review by crodriguezvega — COMMENTED** [review]
Thank you, @angbrav!

I left a bunch of comments for typos or consistency fixes.

Besides that, just mentioning as suggestion maybe the following change:

```diff
-function resultQueryPath(id: Identifier): Path {
+function queryResultPath(id: Identifier): Path {
-   return "queriesresult/{id}"
+   return "queries/{id}/result"
}
```

And I have a couple of questions about the linked list:
- If a querying chain wants to subscribe to a stream from a queried chain, the queried chain must first create this linked list on its side, right?
- If the linked list needs to be created on the queried chain, does that mean that a value that would be queried with a one shot query or streamed needs to be stored twice in the chain? I mean: it needs to be stored in the regular KV store of the chain and besides it also needs to be stored in the linked list to keep track of the previous values it had. Is that correct?

**Comment by angbrav** [comment]
Thank you all for the comments. I've talked to @AdityaSripal and we will work on making this more generally useful primitive.

**Comment by crodriguezvega** [comment]
Agreed with @angbrav to close this PR since the design will take the alternative direction described in #889.

**Pull request closed without merging**